bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/tools/WaitForTopicValue.java

```diff
@@ -24,14 +25,15 @@
 import org.eclipse.smarthome.io.transport.mqtt.MqttMessageSubscriber;
 
 /**
- * Waits for a topic value to appear on a MQTT topic. One-time useable only per instance.
+ * Waits for a topic value to appear on a MQTT topic. One-time usable only per instance.
  *
  * @author David Graeff - Initial contribution
  */
 @NonNullByDefault
 public class WaitForTopicValue {
-    private CompletableFuture<String> future = new CompletableFuture<>();
-    private final CompletableFuture<Boolean> subscripeFuture;
+    private final CompletableFuture<String> future = new CompletableFuture<>();
```

Comment:
I suspect that you should replace usages of `future` in this class with `composeFuture`. 
Your `timeout ` method completes the `future` but the result of that future isn't available until the `subscribeFuture` completes first so your timeout wouldn't actually cause the timeout as you would expect.

However if you caused the timeout on the `composeFuture` everything would operate as you would expect.

Reply:
Thanks @cpmeister, although we're going to change more lines than the class itself has :smile:  Anyway, it seems that the Homie discovery process has another unnoticed error. The Homie300Discovery class does the following:

```java
try {
            WaitForTopicValue w = new WaitForTopicValue(connection, topic.replace("$homie", "$name"));
            w.waitForTopicValueAsync(scheduler, 700).thenAccept(name -> {
                publishDevice(connectionBridge, connection, deviceID, topic, name);
            });
        } catch (InterruptedException | ExecutionException ignored) {
            // The name is nice to have, but not required, use deviceId as fallback
            publishDevice(connectionBridge, connection, deviceID, topic, deviceID);
        }
```
But, if I intentionally remove the $name message and thus, trigger the timeout, the ```composeFuture.completeExceptionally``` function doesn't throw any error, so no discoveryResult is published. I'm sorry, but do you know how can I receive the ```completeExceptionally``` message in the ```w.waitForTopicValueAsync()``` sequence?

This CompleteFuture topic is driving me crazy, I'm more used to synchronous programming :disappointed:

Reply:
Awesome, it works! It seems that I need to read more about ```whenComplete```, ```exceptionally```, ```thenApply```... :wink: